src/logika/Grid.java

```diff
@@ -253,17 +257,89 @@ public List<Index> freeFields() {
 		return free;
 	}
 
-	public void print() {
+    public int minimumNumberOfLiberties(FieldColor color) {
+		int cnt = height*width;
+		for (int i = 0; i < height; i++) {
+			for (int j = 0; j < width; j++) {
+				if (!grid[i][j].equals(color)) continue;
+				int num = ((LibertiesSearchData)libertiesSearch.data).numberOfLiberties(connectedComponents[i][j]);
+				cnt = Math.min(cnt, num);
+			}
+		}
+		return cnt;
+    }
+
+    @Override
+    public String toString() {
+		String buf = "";
 		for (int i = 0; i < height; i++) {
 			for (int j = 0; j < width; j++) {
 				char c = switch(grid[i][j]) {
 				case EMPTY -> '.';
 				case WHITE -> 'o';
 				case BLACK -> '#';
 				};
-				System.out.print(c);
+				buf += c;
+			}
+			buf += '\n';
+		}
+		return buf;
+    }
+
+    public void printToFile(String filename) {
+		try (FileWriter writer = new FileWriter(filename)) {
+			writer.write(toString());
+		}
+		catch(Exception e) {}
+    }
+
+    public static Grid readFromFile(String filename, int size) throws FileNotFoundException, IOException {
+        Grid grid = new Grid(size, size);
+		try (FileReader reader = new FileReader(filename)) {
+			for (int i = 0; i < size; i++) {
+				for (int j = 0; j < size; j++) {
+					char c = (char) reader.read();
+					switch(c) {
+					case 'o':
+						grid.grid[i][j] = FieldColor.WHITE;
+						break;
+					case '#':
+						grid.grid[i][j] = FieldColor.BLACK;
+						break;
+					}
+				}
+				reader.read();  // newline
 			}
-			System.out.println();
 		}
+		return grid;
+    }
+
+	public Grid deepcopy() {
+		Grid m = new Grid(height, width);
+		for (int i = 0; i < height; i++) {
+			for (int j = 0; j < width; j++) {
+				m.grid[i][j] = grid[i][j];
+			}
+		}
+		m.componetSearch.runAll();
+		m.libertiesSearch.runAll();
```

Comment:
Tu lahko samo prekopiraš `grid.connectedComponents`, `grid.colorOfField`, in `grid.liberties`, pa ne bo treba preiskat vsega še enkrat. Nisem prepričan, če be imelo bistven vpliv.

Reply:
Prekopirati bi rabil tudi `libertiesSearch`, kar mi ni všeč.